dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,411 @@
+        {
+            return item;
+        }
+
+        return default;
+    }
+}
+
+sealed internal class KustoMemoryRecord
+{
+    public string Key { get; set; }
+    public MemoryRecordMetadata Metadata { get; set; }
+    public Embedding<float> Embedding { get; set; }
+    public DateTime? Timestamp { get; set; }
+
+    public KustoMemoryRecord(MemoryRecord record) : this(record.Key, record.Metadata, record.Embedding, record.Timestamp?.UtcDateTime) { }
+
+    public KustoMemoryRecord(string key, MemoryRecordMetadata metadata, Embedding<float> embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = metadata;
+        this.Embedding = embedding;
+        this.Timestamp = timestamp;
+    }
+
+    public KustoMemoryRecord(string key, string metadata, string? embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = KustoSerializer.DeserializeMetadata(metadata);
+        this.Embedding = KustoSerializer.DeserializeEmbedding(embedding);
+        this.Timestamp = timestamp;
+    }
+
+    public MemoryRecord ToMemoryRecord()
+    {
+        return new MemoryRecord(this.Metadata, this.Embedding, this.Key, this.Timestamp);
+    }
+
+    public void WriteToCsvStream(CsvWriter streamWriter)
+    {
+        var jsonifiedMetadata = KustoSerializer.SerializeMetadata(this.Metadata);
+        var jsonifiedEmbedding = KustoSerializer.SerializeEmbedding(this.Embedding);
+        var isoFormattedDate = this.Timestamp?.ToString("o", CultureInfo.InvariantCulture) ?? string.Empty;
+
+        streamWriter.WriteField(this.Key);
+        streamWriter.WriteField(jsonifiedMetadata);
+        streamWriter.WriteField(jsonifiedEmbedding);
+        streamWriter.WriteField(isoFormattedDate);
+        streamWriter.CompleteRecord();
+    }
+}
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore"/> backed by a Kusto database.
+/// </summary>
+/// <remarks>The embedded data is saved to the Kusto database specified in the constructor.
+/// Similarity search capability is provided through a cosine similarity function (added on first collection creation). Use Kusto's "Table" to implement "Collection".
+/// </remarks>
+public class KustoMemoryStore : IMemoryStore
+{
+    private string _database;
+    private static ClientRequestProperties GetClientRequestProperties() => new()
+    {
+        Application = "SemanticKernel",
+    };
+
+    private readonly ICslQueryProvider _queryClient;
+    private readonly ICslAdminProvider _adminClient;
+
+    private static readonly ColumnSchema[] s_collectionColumns = new ColumnSchema[]
+    {
+        new ColumnSchema("Key", typeof(string).FullName),
+        new ColumnSchema("Metadata", typeof(object).FullName),
+        new ColumnSchema("Embedding", typeof(object).FullName),
+        new ColumnSchema("Timestamp", typeof(DateTime).FullName),
+    };
+
+    private const string c_collectionPrefix = "sk_memory_";
```

Comment:
I've seen other connectors prefix tables the same way. It is helpful to distinguish between tables managed by an external library vs internally in Kusto.

Is this a hard no for you or can we keep the prefix? I'll change according to your input.

Reply:
I should note that without a prefix, I won't have an easy way to list the collections in the database. I'd have to resort to matching the schema, which I feel is less intuitive.

Reply:
> I've seen other connectors prefix tables the same way.

It was previously, currently we don't have table prefixing in other connectors. If it's the case, please do let me know, since we will need to remove prefixing in these connectors as well.

> It is helpful to distinguish between tables managed by an external library vs internally in Kusto.

The end users should be responsible for naming tables/collections in a way that works for them. It's similar scenario as in Entity Framework. If you use SQL Server for example, Entity Framework (as an external library) doesn't label tables during creation process.


> Is this a hard no for you or can we keep the prefix?

This question was discussed already with internal team and also with open-source community. We decided to remove prefixing from all the connectors.



> I won't have an easy way to list the collections in the database. I'd have to resort to matching the schema, which I feel is less intuitive.

I think you can list all of them. The end user is responsible for DB architecture, so if user inserts record in unmatched schema, that is the same type of problem when using some ORM with DB - it will throw an exception.



> it would force me to check for similarity function existence

That's why you probably need to move `InitializeVectorFunctionsAsync` from `CreateCollectionAsync` to some other place. I think constructor should be a good one. It will be called only once during initialization and used for passed `database` parameter.

Reply:
It is usually considered a bad practice to put initialization code in constructors. 
But, if that is the convention, I'll follow along 😃

Reply:
> I think you can list all of them. The end user is responsible for DB architecture, so if user inserts record in unmatched schema, that is the same type of problem when using some ORM with DB - it will throw an exception.

I somewhat disagree with that statement, because it makes the memory connector non-standalone. Kusto is often used in a way where databases are shared, and contain large number of tables. Creating a database requires elevated permissions, so if a user was to try out the connector on an existing database, it would require them to be _very_ careful in their usage, which is more error prone. 

But again, I'm mostly stating the Kusto perspective. I will align with Semantic Kernel preferences.

Reply:
Agreed in other communication channel to put initialization logic in `GetNearestMatchesAsync` method with additional flag to avoid calling initialization logic every time before Search operation. @danield137 thanks again! :)